contract-message-test/tests/harness.rs

```diff
@@ -1,184 +1,366 @@
-use fuel_crypto::Hasher;
-use fuels::contract::script::Script;
-use fuels::prelude::*;
-use fuels::test_helpers::Config;
-use fuels::tx::{AssetId, Bytes32, Contract as tx_contract, Input, Output, Transaction, UtxoId};
-
-abigen!(
-    TestContract,
-    "../contract-message-test/out/debug/contract_message_test-abi.json"
-);
-
-async fn get_balance(provider: &Provider, address: Address, asset: AssetId) -> u64 {
-    let balance = provider.get_asset_balance(&address, asset).await.unwrap();
-    balance
+mod utils {
+    pub mod builder;
+    pub mod environment;
 }
+use utils::builder;
 
-#[tokio::test]
-async fn spend_predicate_with_script_constraint() {
-    // Set up a wallet
-    let native_asset: AssetId = Default::default();
-    let mut provider_config = Config::local_node();
-    provider_config.predicates = true; // predicates are currently disabled by default
-    let wallets = launch_custom_provider_and_get_wallets(
-        WalletsConfig::new_single(None, None),
-        Some(provider_config),
-    )
-    .await;
-    let wallet = wallets[0].clone();
-
-    // Get provider and client
-    let provider = wallet.get_provider().unwrap();
-    let client = &provider.client;
-
-    // Get padded bytecode root that must be hardcoded into the predicate to constrain the spending transaction
-    let mut script_bytecode =
-        std::fs::read("../contract-message-script/out/debug/contract_message_script.bin")
-            .unwrap()
-            .to_vec();
-    let padding = script_bytecode.len() % 8;
-    let script_bytecode_unpadded = script_bytecode.clone();
-    script_bytecode.append(&mut vec![0; padding]);
-    let script_hash = Hasher::hash(&script_bytecode);
-
-    println!("Padded script length: {}", script_bytecode.len());
-    println!("Padded script hash   : 0x{:?}", script_hash);
-
-    // Deploy test contract
-    let test_contract_id = Contract::deploy(
-        "../contract-message-test/out/debug/contract_message_test.bin",
-        &wallet,
-        TxParameters::default(),
-        StorageConfiguration::default(),
-    )
-    .await
-    .unwrap();
-    let contract_instance = TestContract::new(test_contract_id.to_string(), wallet.clone());
-    println!("Test contract_id   : 0x{:?}", test_contract_id);
-
-    // Get predicate bytecode and root
-    let predicate_bytecode =
-        std::fs::read("../contract-message-predicate/out/debug/contract_message_predicate.bin")
+// Test that input messages can be relayed to a contract
+// and that the contract can successfully parse the message data
+mod success {
```

Comment:
I'm unsure about this test. On one hand you want to test that a single call succeeds and thus you check for all the values, as below, but on the other hand you want to know where exactly something may have failed so you'd factor out common setup and each getter would be tested individually. If one test fails then it's clear where the fault is rather than parsing a larger test. I'm leaning towards the latter with proper utility functions performing setup since this removes duplication.

Reply:
I'm not sure I follow. You're thinking we make a separate test for each `test_contract.get_test_<item>` getter individually? If so, I think they should all be tested together as one big data parsing test. I wouldn't consider just one data piece parsing correctly as a successful test if one of the others didn't.

Reply:
I think @Braqzen is thinking in terms of debugging here. You would still want all the individual data-parsing tests to pass as a whole, but if a single piece fails it's easy to pinpoint exactly where the failure is.
That said, the same can be accomplished with multiple assertions in a single test.